udata/core/dataservices/models.py

```diff
@@ -29,6 +31,42 @@ def visible(self):
     def hidden(self):
         return self(db.Q(private=True) | db.Q(deleted_at__ne=None) | db.Q(archived_at__ne=None))
 
+    def filter_by_dataset_pagination(self, datasets, page):
+        """Paginate the dataservices on the datasets provided.
+
+        This is a workaround, used (at least) in the catalogs for sites and organizations.
+        We paginate those kinda weirdly, on their datasets. So a given organization or site
+        catalog will only list a `page_size` number of datasets, but we'd still want to display
+        the site's or org's dataservices.
+        We can't "double paginate", so instead:
+        - only if it's the first page, list all the dataservices that serve no dataset
+        - list all the dataservices that serve the datasets in this page
+        """
+        # We need to add Dataservice to the catalog.
+        # In the best world, we want:
+        # - Keep the correct number of datasets on the page (if the requested page size is 100, we should have 100 datasets)
+        # - Have simple MongoDB queries
+        # - Do not duplicate the datasets (each dataset is present once in the catalog)
+        # - Do not duplicate the dataservices (each dataservice is present once in the catalog)
+        # - Every referenced dataset for one dataservices present on the page (hard to do)
+        #
+        # Multiple solutions are possible but none check all the constraints.
+        # The selected one is to put all the dataservices referencing at least one of the dataset on
+        # the page at the end of it. It means dataservices could be duplicated (present on multiple pages)
+        # and these dataservices may referenced some datasets not present in the current page. It's working
+        # if somebody is doing the same thing as us (keeping the list of all the datasets IDs for the entire catalog then
+        # listing all dataservices in a second pass)
+        # Another option is to do some tricky Mongo requests to order/group datasets by their presence in some dataservices but
+        # it could be really hard to do with a n..n relation.
+        # Let's keep this solution simple right now and iterate on it in the future.
+        dataservices_filter = Q(datasets__in=[d.id for d in datasets])
+
+        # On the first page, add all dataservices without datasets
+        if page == 1:
+            dataservices_filter = dataservices_filter | Q(datasets__size=0)
```

Comment:
Not introduced in this PR, but dataservice.dataset may not exist (and not be an empty list).
I think we need to add `| Q(datasets__exists=False)`.

Reply:
The final result would be:
```python
        if page == 1:
            dataservices_filter = (
                dataservices_filter | Q(datasets__size=0) | Q(datasets__exists=False)
            )
```

Is that what you were thinking?

Reply:
I don't know why `dataservice.datasets` could be `None`. For me it's always a list… If it's not wanted we could add a migration to convert all `None` to `[]`. If it's wanted maybe add a test with a `dataservice.datasets = None` to be sure this new condition `Q(datasets__exists=False)` is working as intended (and that someone doesn't remove it in 6 months thinking it's weird to have both).

Reply:
From a few tests I made locally, it does seem that even when a Dataservice is created with no datasets (or with its `datasets` set to `None`), the model still ends up with an empty list in `Dataservice.datasets`, probably because the default value of `MongoEngine.ListField.default` is `[]`.

Also, if we wanted to allow non existing datasets, then we'd fail on [this line in udata/core/dataservices/rdf.py](https://github.com/opendatateam/udata/blob/730426458014867b8a7c247dc088b87f876379c8/udata/core/dataservices/rdf.py#L126).

So i'll just revert the change adding ` | Q(datasets__exists=False)` as it seems useless.

@maudetes if you feel we should change this behavior in the future, I think we should address it in a separate issue?

Reply:
I haven't delved into, but we somehow have None value in prod instead of empty list on this dataservice https://www.data.gouv.fr/fr/dataservices/6661e09c7230a8fc3b1ec229/.

We've already faced this situation before, ex: https://github.com/datagouv/data.gouv.fr/issues/1088.
Agreed for tackling it in a separate issue: https://github.com/datagouv/data.gouv.fr/issues/1430